core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -52,7 +57,7 @@ object Executor {
       arguments: Map[String, InputValue],
       path: List[Either[String, Int]]
     ): ReducedStep[R] =
-      step match {
+      transformer.transformStep.lift(step).getOrElse(step) match {
```

Comment:
If I'm understanding this correctly, we will be running the transformers on each `ObjectStep` prior to determining whether it'll be used, is that correct? I'm kind of worried about the runtime performance this might have on large schemas.

I'm still trying to wrap my head around how all this fits together, but I don't suppose it would be possible to apply the transformations during `caliban.execution.Field.apply`? Might be missing something here, but the benefit I see by doing that (assuming it's possible) is that is that we would only be transforming fields that are used as part of the query.

Reply:
Transforming `Field.apply` would only modify the client request, but we need to change the step itself. A very simple example is if the client request `__typename` and you renamed that type. The name will come from the `ObjectStep`.

But your point about performance is very valid.

Reply:
I could transform all the steps tree in advance and save it but the problem is in the `Extend` transformer I made, I also need access to `Field` when I transform the step 🤔 Need to think about this.

Reply:
Maybe I can add `transformQuery` for my specific use case that uses `Field` and remove it from `transformStep`, then I can transform the steps in advance and keep the transformed value.

Reply:
Ah even if I do it in advance it will run every time as soon as it's inside a `QueryStep` or `FunctionStep`...

Reply:
>Transforming Field.apply would only modify the client request, but we need to change the step itself. A very simple example is if the client request __typename and you renamed that type. The name will come from the ObjectStep.

Agh fair point, didn't think of that issue.

Just thinking out loud: I know you're going to work on a refactoring of how we build the graph, but in the meantime perhaps one way that we could avoid the performance overhead would be by having the transformers have some internal caching on whether they are known to transform / not transform a Step. This would require us however to have a way to assign a consistent hash to a step that doesn't need to be recalculated each time we're executing a request.

Can you think of a way that we could achieve that or would that be too difficult / not worth it given you're planning to refactor the graphql resolution?

Reply:
Actually it doesn't need to be unique per se. It can be the hash code of the schema that created the Step perhaps?

Reply:
It would still not work when it's depending on the field. Actually as you can see in the group chat I'm thinking of a completely different approach for handling the gateway feature because it is starting to feel wrong to reuse the current executor. If I do I guess transformers would move there too.